src/store.c

```diff
@@ -0,0 +1,153 @@
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#include "utils.h"
+#include "error.h"
+#include "header.h"
+#include "store.h"
+#include "errno.h"
+
+#define HSK_HEADERS_BIN "/tmp/hnsd.bin";
+
+int
+hsk_store_init(hsk_store_t *store, const uv_loop_t *loop) {
+  store->loop = (uv_loop_t *)loop;
+  store->fd = -1;
+  store->init = false;
+  store->map = NULL;
+  store->location = HSK_HEADERS_BIN;
+  store->headers = malloc(sizeof(hsk_store_t));
+  if (!store->headers)
+    return HSK_ENOMEM;
+  store->timer = NULL;
+  return HSK_SUCCESS;
+}
+
+int hsk_store_uninit(hsk_store_t *store) {
+  store->map = NULL;
+  store->location = NULL;
+  store->headers = NULL;
+  if (munmap(store->map, store->size) == -1)
+  {
+    close(store->fd);
+    return HSK_EFAILURE;
+  }
+  return HSK_SUCCESS;
+}
+
+static void
+after_timer(uv_timer_t *timer) {
+  hsk_store_t *store = (hsk_store_t *)timer->data;
+  assert(store);
+  hsk_store_timer(store);
+}
+
+static void
+hsk_store_timer(hsk_store_t *store) {
+  hsk_store_sync(store);
+}
+
+int
+hsk_store_open(hsk_store_t *store) {
+  // TODO: hardcoded for 65536 blocks. Should auto re-size if we see more than
+  // 65536 blocks.
+  store->size = 65536 * 236;
```

Comment:
Is this related to mmap limits or this is just temporary number until proper fix?

Reply:
This a temporary number. It _will_ break after block 65536 :cry: